docs/abi.rst

```diff
@@ -0,0 +1,560 @@
+.. _abi:
+
+ABI Support
+===========
+
+TODO: brief description
+
+.. warning::
+  ABI support is still taking shape and is subject to backwards incompatible changes.
+    * Based on feedback, the API and usage patterns are likely to change.
+    * For the following use cases, feel encouraged to rely on abstractions.  Expect a best-effort attempt to minimize backwards incompatible changes along with a migration path.
+
+       * :any:`ABIReturnSubroutine` usage for ABI Application entry point definition.
+
+       * :any:`Router` usage for defining how to route program invocations.
+    * For general purpose :any:`Subroutine` definition usage, use at your own risk.  Based on feedback, the API and usage patterns will change more freely and with less effort to provide migration paths.
+
+TODO: warning about how ABI types are intended to be used at the moment (for standard serialization at the incoming and outgoing layers of contract) - internal use may be extremely inefficient with ABI types
+TODO: warning about how internal representation of ABI types may change over time
+
+Types
+------
+
+The ABI supports a variety of types whose encodings are standardized.
+
+Fundamentals
+~~~~~~~~~~~~
+
+Before we introduce these ABI types, it's important to understand the fundamentals of how PyTeal's ABI type system works.
```

Comment:
the way this is worded ("it's important" without an actual reason) makes me think I can skip this section, but in reality I shouldn't. Maybe something like "We first explain the fundamentals of PyTeal's ABI type system, which includes behavior common to all ABI types."